File: src/04_coaching/llm_client.py
"""Client Ollama Cloud (structured output). Aucune logique métier.

POST https://ollama.com/api/chat avec Authorization: Bearer <OLLAMA_API_KEY>
et `format` = JSON-schema -> renvoie le dict parsé du message. Retries/backoff
sur 429/5xx/timeout, message clair sur clé absente / 401.
"""
from __future__ import annotations

import json
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))  # accès src/
import requests
import riotlib as rl

OLLAMA_URL = "https://ollama.com/api/chat"
_MAX_ATTEMPTS = 4


class LLMError(RuntimeError):
    pass


def generate_json(model: str, system: str, user: str, schema: dict,
                  temperature: float = 0.2, timeout: int = 180) -> dict:
    key = rl.load_env().get("OLLAMA_API_KEY")
    if not key:
        raise LLMError("OLLAMA_API_KEY absente du .env")
    body = {
        "model": model,
        "messages": [{"role": "system", "content": system},
                     {"role": "user", "content": user}],
        "format": schema,
        "stream": False,
        "options": {"temperature": temperature},
    }
    headers = {"Authorization": f"Bearer {key}"}
    last = None
    for attempt in range(_MAX_ATTEMPTS):
        try:
            r = requests.post(OLLAMA_URL, json=body, headers=headers, timeout=timeout)
        except requests.exceptions.RequestException as e:
            last = e
            time.sleep(2 * (attempt + 1))
            continue
        if r.status_code == 401:
            raise LLMError("401 — vérifie OLLAMA_API_KEY")
        if r.status_code == 429 or r.status_code >= 500:
            last = LLMError(f"HTTP {r.status_code}")
            time.sleep(2 * (attempt + 1))
            continue
        r.raise_for_status()
        content = r.json()["message"]["content"]
        return json.loads(content)
    raise LLMError(f"échec après {_MAX_ATTEMPTS} tentatives : {last}")

File: web/cf/src/llm_client.ts
/** Client Ollama Cloud structured output — portage de src/04_coaching/llm_client.py. */

export class LLMError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LLMError";
  }
}

export interface GenerateOpts {
  apiKey: string;
  temperature?: number;
  timeoutMs?: number;
  fetchImpl?: typeof fetch;
  sleepImpl?: (milliseconds: number) => Promise<void>;
}

const MAX_ATTEMPTS = 4;

export async function generateJson(
  model: string,
  system: string,
  user: string,
  schema: unknown,
  opts: GenerateOpts,
): Promise<Record<string, unknown>> {
  const fetchImpl = opts.fetchImpl ?? fetch;
  const sleep = opts.sleepImpl
    ?? ((milliseconds: number) => new Promise<void>((resolve) => setTimeout(resolve, milliseconds)));
  const temperature = opts.temperature ?? 0.2;
  const timeoutMs = opts.timeoutMs ?? 180_000;

  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt += 1) {
    try {
      const response = await fetchImpl("https://ollama.com/api/chat", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${opts.apiKey}`,
        },
        body: JSON.stringify({
          model,
          messages: [
            { role: "system", content: system },
            { role: "user", content: user },
          ],
          format: schema,
          stream: false,
          options: { temperature },
        }),
        signal: AbortSignal.timeout(timeoutMs),
      });
      if (response.status !== 429 && response.status < 500 && !response.ok) {
        throw new LLMError(`ollama HTTP ${response.status} (auth/requête invalide)`);
      }
      if (response.ok) {
        const body = await response.json() as { message?: { content?: string } };
        try {
          const parsed = JSON.parse(body.message?.content ?? "") as unknown;
          if (typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)) {
            return parsed as Record<string, unknown>;
          }
        } catch {
          // Une réponse JSON Ollama dont le contenu n'est pas JSON est retentée.
        }
      }
    } catch (error) {
      if (error instanceof LLMError) throw error;
      // Timeout, erreur réseau et réponse Ollama non JSON sont retentés.
    }
    if (attempt < MAX_ATTEMPTS - 1) await sleep(2_000 * (attempt + 1));
  }
  throw new LLMError(`ollama : échec après ${MAX_ATTEMPTS} tentatives`);
}
